xarray/core/datatree_mapping.py

```diff
@@ -1,236 +1,129 @@
-                )
-            )
-            func_with_error_context = _handle_errors_with_path_context(
-                node_of_first_tree.path
-            )(func)
-
-            if node_of_first_tree.has_data:
-                # call func on the data in this particular set of corresponding nodes
-                results = func_with_error_context(
-                    *node_args_as_datasetviews, **node_kwargs_as_datasetviews
-                )
-            elif node_of_first_tree.has_attrs:
-                # propagate attrs
-                results = node_of_first_tree.dataset
-            else:
-                # nothing to propagate so use fastpath to create empty node in new tree
-                results = None
-
-            # TODO implement mapping over multiple trees in-place using if conditions from here on?
-            out_data_objects[node_of_first_tree.path] = results
-
-        # Find out how many return values we received
-        num_return_values = _check_all_return_values(out_data_objects)
-
-        # Reconstruct 1+ subtrees from the dict of results, by filling in all nodes of all result trees
-        original_root_path = first_tree.path
-        result_trees = []
-        for i in range(num_return_values):
-            out_tree_contents = {}
-            for n in first_tree.subtree:
-                p = n.path
-                if p in out_data_objects.keys():
-                    if isinstance(out_data_objects[p], tuple):
-                        output_node_data = out_data_objects[p][i]
-                    else:
-                        output_node_data = out_data_objects[p]
-                else:
-                    output_node_data = None
-
-                # Discard parentage so that new trees don't include parents of input nodes
-                relative_path = str(NodePath(p).relative_to(original_root_path))
-                relative_path = "/" if relative_path == "." else relative_path
-                out_tree_contents[relative_path] = output_node_data
-
-            new_tree = DataTree.from_dict(
-                out_tree_contents,
-                name=first_tree.name,
-            )
-            result_trees.append(new_tree)
-
-        # If only one result then don't wrap it in a tuple
-        if len(result_trees) == 1:
-            return result_trees[0]
-        else:
-            return tuple(result_trees)
-
-    return _map_over_datasets
+    from xarray.core.datatree import DataTree
+
+    # Walk all trees simultaneously, applying func to all nodes that lie in same position in different trees
+    # We don't know which arguments are DataTrees so we zip all arguments together as iterables
+    # Store tuples of results in a dict because we don't yet know how many trees we need to rebuild to return
+    out_data_objects: dict[str, Dataset | None | tuple[Dataset | None, ...]] = {}
+
+    tree_args = [arg for arg in args if isinstance(arg, DataTree)]
+    name = result_name(tree_args)
```

Comment:
```suggestion
    tree_args = [arg for arg in args if isinstance(arg, DataTree)]
    
    if tree_args == []:
       raise ValueError("Must pass at least one `DataTree`")
    
    name = result_name(tree_args)
```

Reply:
You already get an error message like this (which is tested):
```
>>> xr.map_over_datasets(lambda x: x, "dt")
TypeError: must pass at least one tree object
```